### PR TITLE
Support for NDArray multiplication

### PR DESCRIPTION
**Pull request by cako**
This PR takes over from #353 in introducing valid multiplication by ND-arrays without the need for reshaping
The current version works like this:

1. Every LOp has `dims` and `dimsd`. They are arbitrary tuples which must be compatible with LOp.shape.
2. Suppose that LOp.dims = (5, 10) and LOp.dimsd = (9, 10). The following shapes are obtained on LOp @ x

   |    | Reference   | x.shape                 | (LOp @ x).shape                           | Note                                             |
   |----|-------------|-------------------------|-------------------------------------------|--------------------------------------------------|
   | V0 | (50,)       | (90,)                   | Standard vector multiplication            |                                                  |
   | V1 | (5, 10)     | (9, 10)                 | "Vector" of size (5, 10)                  |                                                  |
   | M0 | (50, 1)     | (90, 1)                 | Standard one-column matrix multiplication |                                                  |
   | M1 | (5, 10, 1)  | (9, 10, 1)              | "One-column matrix" of "vector" (5, 10)   |                                                  |
   | M2 | (50, 20)    | (90, 20)                | Standard matrix multiplication            |                                                  |
   | M3 | (5, 10, 20) | (9, 10, 20)             | "Matrix" of 20 x (5, 10)                  |                                                  |
   |    |             | (1000,)                 | error                                     | Could be reshaped to (50, 20) but is ambigous |
   |    |             | any other kind of shape | error                                     |                                                  |
3. SciPy solvers will struggle with V1, M1 and M3. These cases can be coerced to error if LOp.flattened is set to True. Any operator with flattened = True will have legacy behavior. I'm happy to hear about other ways on how we can do this.

This PR also provides a `reshape`` decorator. It relies primarily on the name of the function being decorated to find out which dimensions to reshape by, and also includes an optional swapaxis call. See the new Symmetrize operator for examples.

### CODE REVIEW

**Comment by mrava87** [comment]
Great, this is a nice addition! As I previously said, if this can be done is great as long as it doesn't make things confusing on the user side (for those users that don't care about this), and it seems like this is the case :)

I have a number of questions/comments:

1. You say `SciPy solvers will struggle with V1, M1 and M3.`. To the best of my knowledge scipy solvers can only handle single RHS (ie. 1d `b` vectors), and the documentation of eg [cg](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.cg.html) says that. I tried just to be sure:2. 

```
from scipy.sparse.linalg import cg as sp_cg

ny, nx, nrhs = 100, 50, 4
x = np.outer(np.ones(nx), np.arange(nrhs)+1)
A = np.random.normal(0, 1, (ny, nx))
A = A.T @ A
Aop = pylops.MatrixMult(A)
y = A @ x

xest = sp_cg(Aop, y, maxiter=4)

```
and as expected this breaks. So also M2 won't work. Only V0 (and apparently M0 for scipy cg but not for scipy lsqr). 

2. The same applies to our own solvers with the 'lucky' exception of ISTA/FISTA (see https://github.com/PyLops/pylops/issues/264). This is more by chance than by design, because ISTA/FISTA never requires computing a vector norm. Instead for any of the solvers in `pylons.optimization.solver` submodule, multiple RHS are not (yet) allowed because of this:

```
self.kold = self.ncp.abs(self.r.dot(self.r.conj()))
```

Then we would need to do:

```
self.kold = self.ncp.sum(self.ncp.abs(self.r * self.r.conj()), axis=0)
```

which is not out of scope (we can maybe do it after we have reorganized all the solvers in classes) but not possible today

2.

File: pylops/LinearOperator.py [hunk 1]
```diff
@@ -82,6 +82,7 @@ def __init__(
         dimsd=None,
         clinear=None,
         explicit=None,
+        flattened=None,
```
Comment (mrava87) [inline pylops/LinearOperator.py:85]: Seems like you have not added flattened to the doctoring?

File: pylops/optimization/leastsquares.py [hunk 2]
```diff
@@ -108,6 +108,18 @@ def NormalEquationsInversion(
     """
     ncp = get_array_module(data)
 
+    # Ensure compatibility with SciPy functions
```
Comment (mrava87) [inline pylops/optimization/leastsquares.py:111]: Mmh I am not sure I understand this. If I understand correctly you simply check if the operator is not flattened you apply flattened, so that the solver will simply go from a multiple RHS problem to a flattened problem?

I think this is reasonable but comes with some caveats. Imagine you want to do a N-traces seismic inversion and write it as a single ``PoststackLinearModelling`` operator. You can then include a ``Laplacian`` regularizers to couple those traces. Now let's take a (perhaps stupid) case where you have 2 datasets and want to put them side by side into a N*2 matrix ``y`` and use the new option M2 to invert for a N*2 matrix  ``x``. This would flatten the operators and couple the laplacian on the first and second column of x?

I am just not sure, I am asking as you may have thought about ;)

File: pylops/utils/decorators.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from functools import wraps
+from typing import Callable, Optional
+
+
+def reshaped(
+    func: Optional[Callable] = None,
```
Comment (mrava87) [inline pylops/utils/decorators.py:6]: mmh have we started introducing typing annotations?

I wonder if it is best to wait and then do this in one unique 'boring' PR so we make sure everything is done consistently across the library?

File: pylops/LinearOperator.py (hunk 1 above)
Comment (cako) [reply pylops/LinearOperator.py:96]: Done

File: pylops/optimization/leastsquares.py (hunk 2 above)
Comment (cako) [reply pylops/optimization/leastsquares.py:111]: > Mmh I am not sure I understand this. If I understand correctly you simply check if the operator is not flattened you apply flattened, so that the solver will simply go from a multiple RHS problem to a flattened problem?

Not exactly. [The logic in the flattened part of the LinearOperator](https://github.com/PyLops/pylops/blob/c4abb6a54d5efdcf279a749b6d4acdbf184913fc/pylops/LinearOperator.py#L508-L513) does not flatten N-d array inputs. If the arrays are not shaped (N,) or (N, P), the multiplication will raise an error. This was the same behavior as the v1.x, so it is basically an easy way to ensure that your code doesn't break going from v1.x to v2. I'm tempted to make the default rely on a module-level variable but I haven't decided if that is a good idea or not. Something like `pylops.set_ndarray_multiplication(True/False)`.

In this particular function (leastsquares), I ensure all input operators are flattened = True, and then set them back. It's a bit hacky. The module-level implementation would make the code considerably easier, but globals are usually not a great idea.

File: pylops/utils/decorators.py (hunk 3 above)
Comment (cako) [reply pylops/utils/decorators.py]: Fair enough! Removed them.

File: pylops/optimization/leastsquares.py (hunk 2 above)
Comment (cako) [reply pylops/optimization/leastsquares.py:111]: Just an addition to this. I introduced the flattened logic because of the annoying case when dims is flat i.e., dims = (N,) but dimsd is not e.g. dimsd = (M1, ..., MK). This case appears in interpolation. I previously (erroneously) assumed that flat inputs -> flat outputs. But that is clearly not the case, which means we can't rely on the shape of dims to decide whether to flatten the output or not. 

I figured the best solution was to separate the v2 behavior from v1. In v1 you always get an error when you try n-d array multiplication (unless its a properly shaped matrix). In v2, if we have a dims-shaped input, we will output a dimsd-shaped array. In the case above (dims is flat but dimsd is not), it *will* reshape the output. Taking the interpolation example, this makes sense.

**Comment by cako** [comment]
1. Fair enough, I didn't realize this. But it's ok, we don't use it like that and we don't plan on using it like that.
2. Yea, we at some point need to decide if we want to support the matrix `b`. Regarding the n-d shaped inputs, I didn't notice any issues but maybe to be on the safe side we can also "protect" the operators (set them to flattened)? (Starting to think `pylops.set_nd_multiplication` is not such a bad idea after all 😅

File: pylops/optimization/leastsquares.py (hunk 2 above)
Comment (mrava87) [reply pylops/optimization/leastsquares.py:111]: Mmh I’m confused. If you refer to Interp/Bilinear, I would say the input has dims.. if that is 1D (dims=(N,), how can you get an output that is ND?

Comment (mrava87) [reply pylops/optimization/leastsquares.py:111]: Regarding the first comment, so you put all operators to flattened=True so that their outputs are flat (which is what solvers need at the moment)?

Comment (cako) [reply pylops/optimization/leastsquares.py:111]: > Mmh I’m confused. If you refer to Interp/Bilinear, I would say the input has dims.. if that is 1D (dims=(N,), how can you get an output that is ND?

The issue I was having was with ``Bop.H`` in plot_bilinear.py. It takes a flat array and outputs a 2D array. Maybe I'm doing something wrong?

Comment (cako) [reply pylops/optimization/leastsquares.py:111]: > Regarding the first comment, so you put all operators to flattened=True so that their outputs are flat (which is what solvers need at the moment)?

Exactly. They will also error if the input is not flat (v1.x behavior).

Comment (mrava87) [reply pylops/optimization/leastsquares.py:111]: > > Mmh I’m confused. If you refer to Interp/Bilinear, I would say the input has dims.. if that is 1D (dims=(N,), how can you get an output that is ND?
> 
> The issue I was having was with `Bop.H` in plot_bilinear.py. It takes a flat array and outputs a 2D array. Maybe I'm doing something wrong?

That's correct. I was referring to the forward and that one goes from grid to random points, so the adjoint goes from random points to a grid :)

**Comment by mrava87** [comment]
> * Yea, we at some point need to decide if we want to support the matrix `b`. Regarding the n-d shaped inputs, I didn't notice any issues but maybe to be on the safe side we can also "protect" the operators (set them to flattened)? (Starting to think `pylops.set_nd_multiplication` is not such a bad idea after all 😅

I think ultimately it is a good idea, as you can leverage as much as possible vectorization to apply the operators... it just needs some changes in the solver, which we can plan to do. What would `pylops.set_nd_multiplication` do?

**Comment by cako** [comment]
My idea was to use this set_nd_multiplication method as a global way of turning the ND-array multiplication on/off. I don't think it actually makes sense to keep it on a per-operator basis as I have implemented it.

**Comment by mrava87** [comment]
Alright I see. So that if it is turned off the matvec/rmatvec will always flattens the input arrays before passing them to the internal methods (_matvec/_rmatvec) or they will throw an error if an nd-array is passed?

**Comment by cako** [comment]
It should throw an error (v1.x behavior)

File: pylops/optimization/leastsquares.py (hunk 2 above)
Comment (cako) [reply pylops/optimization/leastsquares.py]: I just committed the changes for setting the ndarray support on a module level. This also creates a nice pattern to alter behavior within functions for backward compatibility like this:

```python
from pylops import disabled_ndarray_multiplication

def old_style_func(Op, y, ...):
    x = x.ravel() # Optional, otherwise will error on non-1d arrays
    with disabled_ndarray_multiplication():
        # do stuff in v1.x way
```

Let me know if that makes sense. If you do, just resolve the conversation :)

**Comment by cako** [comment]
Hey @mrava87, I'm pretty happy with how it works now. I'm tempted to merge with dev before Transform 2022 if you are happy as well!

File: pylops/optimization/leastsquares.py [hunk 4]
```diff
@@ -106,65 +107,66 @@ def NormalEquationsInversion(
     implicitly assumed to be zero.
 
     """
-    ncp = get_array_module(data)
+    with disabled_ndarray_multiplication():
```
Comment (mrava87) [inline pylops/optimization/leastsquares.py:110]: Why is this needed here? And not in other solvers in this module?

**Comment by mrava87** [comment]
This looks good :) 

Just to be sure, I have two minor questions:

- you have so far created a framework for handling NDArray multiplications and changed only one operator to use the reshaped decorator (Symmetrize). I guess you will make a new PR following this for all the other operators where this code pattern make sense?

- you have flagged this as breaking. Is it really or by default this would not change any behaviors of v1. I ask because you have not added anything to the MIGRATION_V1_V2.md file. Have a think about if anything is needed there.

I think you can merge once you are happy with it.

File: pylops/optimization/leastsquares.py (hunk 4 above)
Comment (cako) [reply pylops/optimization/leastsquares.py:110]: I was having an issue with this solver before because it calls the SciPy cg which does not accept NDarray multiplication. However, I looked into it a bit deeper and it seems that there are several instances in the codebase where methods that only work for 1d arrays are called (e.g. scipy lsqr and cg). They are mostly internal, and since out tests are all built for v1.x behavior, I didn't catch this.

One solution is to create a wrapper for these scipy functions so that users can understand how to interact with functions that might be reliant on v1.x behavior (but that they cannot change/are external). It would work something like this:

```python
# v1.x code: only works with 1d arrays
from scipy.sparse.linalg import lsqr
xlsqr, *_ = lsqr(Sop, y.ravel())

# proposed new code
def add_ndarray_support_scipy(func): # this function would be in pylops.utils.decorators
    """
    Decorator which converts a SciPy-solver-type 
    function into one that supports ndarray multiplication.
    """
    def wrapper(A, b, *args, **kwargs): # this is the signature of sparse solvers in scipy
        with disabled_ndarray_multiplication():
            res = func(A, b.ravel(), *args, **kwargs)
            res[0] = res[0].reshape(A.dims)
            return res

from scipy.sparse.linalg import lsqr
lsqr = add_ndarray_support_scipy(lsqr) # "monkey-patch"
xlsqr, *_ = lsqr(Sop, y) # works with ndarrays

```

**Comment by cako** [comment]
1. Yes, I think the best course of action is to refactor the operators not only to use the best practices from this PR, but also from some previous ones. The main changes that would be done are: do not register shape or dims/dimsd directly, instead call `super().__init__(dims=dims, dimsd=dimsd)`.
2. I flagged this as breaking mostly because of some issues I had while running the tests after I implemented the this functionality. Yes, code that worked before *should* keep working, but there were a couple of places that had bugs that needed to be fixed before the change was actually non-breaking. Not sure if this is really "breaking", but better safe than sorry :) With regard to the migration, I will fill that out!

File: pylops/optimization/leastsquares.py (hunk 4 above)
Comment (mrava87) [reply pylops/optimization/leastsquares.py:110]: Mmh, I see. Yes we have quite a few functions that call a solver, and if the input is a numpy array by default they would call the appropriate scipy solver  (already if it is cupy they will call our equivalent as at that time cupy did not have scipy solvers in place for linear operators.. I need to check if they do now). 

We could to begin with do this, and in the longer run when all our solvers are refactored change them so that they can handle multiple rhs, then check if this is the case and use them instead of scipy ones… or even switch all together to our solvers and just have some tutorial that shows you can use scipy solvers but we now reccomend using ours?

**Comment by mrava87** [comment]
Btw, not sure if you want to do it now with Symmetrize or later with another operator, but I think having an example that shows how to use these new functionalities would be valuable, otherwise they may end up being ‘undiscovered’ ;)

File: pylops/optimization/leastsquares.py (hunk 4 above)
Comment (cako) [reply pylops/optimization/leastsquares.py]: I implemented the above method (with some small alterations) and in general it works. I also added in the migration readme how to use operators with the new LinearOperator (either disable nd-array mult or decorate them).

The only thing missing are the sparse solvers, but I need to look into those a bit better. For now the solution to those is to disable nd-array support when calling them.

**Comment by cako** [comment]
I added the NMO example, but I want to go through all the examples/tutorials and sort everything out. I'll merge before T22 so we can get more eyes on this!

**Comment by mrava87** [comment]
Sounds good. For the NormalEquationInversion it is ok, but please do not spend time on sparsity as I plan to have all operators in class based mode (or at least a first draft) in the next week or so. So it’s better we wait and then see based on that what problems we get with the new Ndarray options and decide which changes to make. I NormalEquationInversion is actually ready for me to push. I will look at what you changed and include it there as well as for other solvers

**Comment by cako** [comment]
Awesome, let's wrap up the solvers first. W.r.t the NormalEquationInversion, very few changes so it should be easy to make it work :)
